File: Python/unique-email-addresses.py
# Time:  O(n * l)
# Space: O(n * l)

class Solution(object):
    def numUniqueEmails(self, emails):
        """
        :type emails: List[str]
        :rtype: int
        """
        def convert(email):
            name, domain = email.split('@')
            name = name[:name.index('+')]
            return "".join(["".join(name.split(".")), '@', domain])

        lookup = set()
        for email in emails:
            lookup.add(convert(email))
        return len(lookup)
  

File: C++/unique-email-addresses.cpp
// Time:  O(n * l)
// Space: O(n * l)

class Solution {
public:
    int numUniqueEmails(vector<string>& emails) {
        unordered_set<string> result;
        for (const auto& email : emails) {
            result.emplace(convert(email));
        }   
        return result.size();
    }

private:
    string convert(const string& email) {
        const auto& at_it = email.find('@');
        const auto& domain = email.substr(at_it);
        auto name = email.substr(0, at_it);
        name = name.substr(0, name.find('+'));
        name.erase(remove(name.begin(), name.end(), '.'), name.end());
        auto new_email = move(name);
        new_email += domain;
        return new_email;
    }
};
